### ISSUE TITLE
Implement Preference Setting(s) for: 'Preferred "Save to" Device', and/or a(n) (Never-Save-To) 'Exclusion List'

### ISSUE THREAD

**Issue by schroeder82** [Ultimaker/Cura#16472]
### Broad overview of 'issue' that sparks me desire for an additional set of features:

I have 1 flash drive that is used as a general backup drive on my computer, used to save and store various random files. This drive is _always_ plugged into my computer and rarely gets removed.

My '**PRINTER**' Flash drive, which I use exclusively for saving my G-Code onto, is most commonly plugged into my QIDI X-Plus printer. As I open/load a .STL file via CURA, I go to my printer and start heating it up; It's at this time that I grab the '**PRINTER**' drive, that resides in my printer, and place it into my PC. {_Doing this allows the file/part(s)/program to load while I start my printer, **and** then allows my printer to pre-heat while I return to my computer to tweak any Slicer settings before Exporting the G-Code_}

It's always been a frustration that (_as far as my limited knowledge_): I'm not able to setup CURA to [**Automatically**] default to a _specific_ Flash Drive. Thus, There have been many a time in which: I click to **{Save to Removable Drive (Whatever _Drive Letter_ has been defaulted:)}**, Eject or Unplug the '**PRINTER**' drive, Plug it into my Printer, only to realize CURA saved, _by default_, to my other flash drive that is used for backups, rather than my drive intended exclusively for printer G-Code.

---------------------

I'm aware that: I can _manually_ select the _correct_ flash drive when saving; But that's often a forgotten action in the excitement of trying to get my print started. I currently subvert this '_issue_' by manually ejecting my **'Storage/Backup**' drive (whenever I expect to be using CURA), so that the '**PRINTER**' drive is the exclusive option that can even be seen. This '_Works_', until my computer needs to be restarted for any reason; If my computer is restarted then my '**Storage/Backup**' USB is reassigned a drive letter and is visible, requiring me to eject it again or realize my oversight when I inevitably save my G-Code to that '**Storage/Backup**' drive and can't find it when trying to locate the file on my printer's menu, when searching through my '**PRINTER**' drive.

### A minimalized solution would be having the ability to select a default _Drive Letter_ as a '**Preferred 'Save to' Device**'.

>"**Device**" might not be the most ideal term, but it's intended to be a term that conveys: Disks, Removable Drives, SD Cards(?), and any other storage hardware '_Devices_'. {_I've never saved to an SD Card from CURA, so it might already be considered a Removable Drive, just like a USB Flash Drive._}

---------------------

### Example of implementation:
>You navigate to **CURA**->[**Preferences**]->[**Configure Cura...**]       (_The **"Preferences" Form** Opens_)
Under the [**General**] Tab, "**Opening and saving files**" section, there would be a new entry labeled '**Preferred 'Save to' Device**'.
>
>'**Preferred 'Save to' Device**' would have a dropdown-style Listbox UI object, which, by default, is set to "**_None_**". You click the Listbox and scroll to whatever _Drive Letter_ is associated with your preferred location (ex. "**Z:**"). 
>
>Directly below the setting for '**Preferred 'Save to' Device**', would be a checkbox option labeled '**Automatically override "Save to _{Location}_" to Selected Preference, whenever possible.**'

---------------------

### How this would work/help, for my use-case.

This '_Solution_' would require you to, firstly, assign a _**Persisting** Drive Letter_ within your operating system (A feature that's at least available within Windows, I'm unfamiliar with other OS's features). For me, I would take my '**PRINTER**' Flash Drive and set up a _**Persisting** Drive Letter_ of "**Z:**" to ensure it's almost impossible that there would be a _drive letter_ collision; Thus, any time my '**PRINTER**' drive is plugged into my computer, it get assigned as "**Z:**". 

I would then be able to have "**Z:**" selected within the **CURA**->[**Preferences**]->[**Configure Cura...**] >>  [**General**]->"**Opening and saving files**"->'**Preferred 'Save to' Device**'. This selection alone would mean that, Upon Loading CURA, if "**Z:**" is available, CURA will automatically set the "**Save to _{Location}_**" (_Button_) mapping to "**Save to REMOVEABLE DRIVE (Z:)**".

>Furthermore, I, _personally_, would check (_Enable_) the option for: **CURA**->[**Preferences**]->[**Configure Cura...**] >> [**General**]->"**Opening and saving files**"->'**Automatically override "Save to _{Location}_" to Selected Preference, whenever possible.**' This would mean that: Regardless of wherever the "**Save to _{Location}_**" (_Button_) _currently_ happens to be mapped, If I connect my preferred '**Preferred 'Save to' Device**', "**Z:**", Then the software would _automatically_ **override** the "**Save to _{Location}_**" (_Button_) mapping to "**Save to REMOVEABLE DRIVE (Z:)**".

---------------------

### Potential issues

This would be the most minimal implementation that I can currently come up with, which would sufficiently resolve my '_issue_' in a satisfactory manner. I believe this would also be reasonably simple to implement; but it's obviously "_easier said than done_" and there may be a plethora of alternative 'Connect via USB' and 'Saving' features, which already exist, that might have interactions with the implementation of _this_ proposed feature.

### _Here are some alternative implementations or additional features that would couple well with the initial idea:_

---------------------
### Additional depth to initial implementation:

As most Printers are, _I presume_, likely to require files to be located on the _Root_ Location of a given Device, I'm not sure on how much of a benefit could be gained by adding an additional layer of depth to this implementation. However, I can see that, for some users, it might be even better if you could add a _Preference of **Folder**_ within a given **Preferred Device**.

So, Beyond just '**Preferred 'Save to' Device**', there would be an Option such as: '**Preferred _Folder_ on Device**'. Setting this _Path_ might be best as _'**Browse Folder-Location**'-style_ of selection. You could click a '**Browse Folder-Location**' (_button_) and use the typical/canned browse to folder means of selection. 

>The downside here is that you would have to have the device connected, _so **that** might present some degree of inconvenience_; This could be handled by pairing a _**Textbox**_ and _**Browse-to-folder selector**_ object. If you know the Folder path but don't have the device connected, you could simply type the path into the textbox: "**Z:\UnprintedFiles**".
>
>An implementation consideration is that: When a person goes to save the G-Code to their '**Preferred Device**' & has a set '**Preferred Folder**', the "**Save to REMOVEABLE DRIVE (Z:\UnprintedFiles)**" _button-click event_ would need to verify that the folder path "**\UnprintedFiles**" actually exists, prior to saving. {_The folder might have been moved, deleted, or never existed._}
>
>CURA could just, _by default_, create the folder and save the file inside the folder, _without having to throw an exception error/prompt_.
>
>>Otherwise, This would mean that an exception should be thrown, and that would imply the need for the simple addition of a checkbox of: **"_Automatically_ Create Folder on '_Preferred Device_' if '_Preferred Folder_' path does not exist".** 

### A little more flexibility

As a potential 'Nice to have', I could see where other people would have a use-case in which they would benefit from an additional degree of control. 

>For example: If you have several printers (ex. _Some that use USB drives and others that require SD cards_), it might be nice to have the ability to: enable more than one '**Preferred Device**'; add them in a listbox; and _click+drag_ to arrange the drives into a **hierarchical order** that would enable the '**Automatically override "Save to _{Location}_" to Selected Preference(s), whenever possible.**' option to set the "**Save to _{Location}_**" (_Button_) to whichever option, _if any are available_, **is highest in priority**.
>
>>I can't really fathom a use-case where each '**Preferred Device**' would need any more than a 1-to-1 mapping for a '**Preferred _Folder_ on Device**'. 

---------------------
---------------------

### A potentially viable alternative

As an alternative implementation, _or potentially an additional feature_: it would be possible to attain a _similar_ functionality, within _my_ use-case, by having an **Exclusion List**. 
>In this regard, you basically add '_Locations_' to a blacklist, which the software will intentionally avoid; in the scope of the initial concept, a '_Locations_' would be a '_**Drive Letter**_'.

This would mean that: I could assign my '**Storage/Backup**' drive to a _**Persisting** Drive Letter_, ex. "**Y:**"; After doing so, I could add multiple locations like "**Y:**" and "**C:**" to the **Exclusion list**. By doing this, the "**Save to _{Location}_**" (_Button_)  would never map to either "**Y:**" or "**C:**"; If I don't have my '**PRINTER**', G-Code exclusive, flash drive plugged into my computer, the "**Save to _{Location}_**" (_Button_)  would then default to something like "**No Save Locations Available**" and _become disabled._ 

>This would effectively ensure that: I would never accidentally save my G-Code to the locations that aren't intended. 

This implementation, of an **Exclusion List**, would likely be effective in _my_ use case, _because I only have 3 drives that are assigned whenever my computer restarts (not including my '**PRINTER**' USB)_ : ex. (**C: ,  L: ,**  and  **E:**), "**L:**" is an additional _internal SSD_. For me, it would be trivial to assign 2 _**Persisting** Drive Letters_ (_One for my '**additional**' SSD and one for my '**Storage/Backup**' USB_) and add them to an **Exclusion List**; but that might not be the same for others. 

>Furthermore, even if an **Exclusion List** were to be implemented **without** the "**Preferred Device/Location**" features, the '_problem/frustration_' would still occur if a user plugs in any set of 2+ '**_temporary_**' drives. One of them might be your _intended_ "**Save to _{Location}_**", but CURA might default to another, _random_, available location, which in this example would be a drive that is only temporarily connected and would never warrant either: _having an assigned, _**Persisting** Drive Letter__ **or** _being added to an **Exclusion List**._

### Who benefits?!?

The inception of this '_set of features_' came from my experience of wanting to save my prints to a **specific** USB drive, _without the hassle that results from also having another USB drive connected to my computer._ 

>The USB I use for '**Storage/Backups**' is a larger drive at **250gb** of capacity, and is formatted in **exFAT** (_out of necessity_). A **FAT32-formatted** drive, like the one I use for '**PRINTING**' is limited to a max of **32gb** of allocation; _my printer can handle the FAT32 format, without issue._
>
>**I would expect that a fair portion of the 3D printers in use, _currently_, would have a drive format limitation.** So anyone who has a printer **and** wants to have a larger, connected storage-device, _as I have_, should benefit from these features.

I recognize that: in the time I've spent (_several hours_) writing out these thoughts: for potential features, and trying to format things, I could've manually clicked the **"Save to ..."** button: to adjust to an alternate location, _based on the current implementation_, thousands of times.

The point is that: If this/these features were added, it might save a **non-trivial** proportion of the Cura users _a couple of seconds per G-Code file save_, **every time**. Cura states, **"Ultimaker Cura users worldwide prepare over 2 million prints per week."**

---------------------

**If we assume the following assumptions and napkin math:**

- There won't be detrimental interactions between these features and pre-existing features, which would extend dev time
- It _might_ take a dev 1 week to implement these features: _**TTI** = Time-To-Implement_
- There are 40hrs in a standard work-week
- 3,600s/hr.
- (1-week/**TTI**) * (40hrs/week) * (3,600s/hr) = 144,000 seconds/**TTI**
- 2,000,000 prints/week
- 2s wasted-time by users per print, who have to adjust save location -> (2s **WT**/print): _**WT** = Wasted-Time_
- (2,000,000 prints/week) * (2s **WT**/print) = (4,000,000s **WT**/week)
- (4,000,000s/144,000s) = 27+(7/9)

So, _if those nominal metrics are even remotely close to accurate_, then **The time wasted**: _selecting an alternative save location_, **for all of the Cura prints _(in single a week)_**, **_could_** equate to about **~27.8x** the amount of time it might take 1 developer **to implement these features**. 

If you take 1/~27.8, we're talking .036; **So, (_Roughly speaking, in theory_) only 3.6% of Cura prints would need to be affected by the same irritation that I experience, before it would seem worth while to have a dev spend ~1 week of time implementing these features.**

>I'm trying to justify the implementation of **all** of these potential features, while only looking at **the value that is likely to be gained by just the inclusion of a 'Preferred Device' option regarding file-saving**; doing so, knowing that these additional features: '_An Exclusion List_', '_Preferred Folder Location_', and '_A "Preferred Device(s)" hierarchy_' would only further extend the reach of how many users would benefit.

Personally, I feel that these '_irritating_' UI/implementation interactions are likely to affect more than 3.6% of prints being saved. Obviously, Cura Devs will have the kind of internal metrics to see how reasonable or unreasonable my nominal values are. **But, again, even if my assumed values are off by an order of magnitude, we're looking at the impact for all of the Cura prints that occur _(in single a week)_.**

### In conclusion:

I feel that: If the internal data shows that **even just a few %** of weekly prints are saved, in circumstances _where **more than one** additional "**Save to ...**" device is available (**beyond the user's OS Disk storage**)_, then these features should be **sincerely** considered and brought forth in discussion for development. 

---------------------

If there's enough interest, I would be willing to spend more time creating a mock-up, simplified GUI in C++ that would help demonstrate the concept and value of this set of features.

**Comment by Slug-Cat** [comment]
Thsi feature would be appreciated. I face the same issues when using more than one flash drive on the same device.
